### ISSUE TITLE
CA1508: False positive in switch expression when conditions

### ISSUE THREAD

**Issue by AnthonyMastrean** [dotnet/roslyn-analyzers#7338]
### Analyzer

**Diagnostic ID**: [CA1508](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Avoid dead conditional code`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.100](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

The analyzer warns on the below code despite no condition that would indicate `value` is not null.

### Steps To Reproduce

I'm having trouble minimizing the example, but the `when` condition seems to be required.

```csharp
static class Ca1508FalsePositive
{
    private static bool Example(string? value) => value switch
    {
        null => false,
        not null when string.IsNullOrEmpty(value) => true, // CA1508
        not null => false, // CA1508
    };
}
```

This doesn't raise the warning.

```
static class Ca1508NoWarnings
{
    private static bool Example(string? value) => value switch
    {
        null => false,
        _ when string.IsNullOrEmpty(value) => true, // CA1508
        _ => false, // CA1508
    };
}
```

### Expected behavior

No warning!

### Actual behavior

Warnings on both `not null` arms of the switch expression.

```
'not null when string.IsNullOrEmpty(value) => true' is always 'true'. Remove or refactor the condition(s) to avoid dead code.
```

```
'not null => false' is always 'true'. Remove or refactor the condition(s) to avoid dead code.
```

## Additional context

Might be related to #5142
